sympy/solvers/ode/ode.py

```diff
@@ -1639,6 +1639,34 @@ def odesimp(ode, eq, func, hint):
     if not isinstance(eq, Equality):
         raise TypeError("eq should be an instance of Equality")
 
+    if eq.lhs.is_Piecewise or eq.rhs.is_Piecewise:
+        # Ignore any simple constraints imposed on the arbitrary constants from evaluating
+        # the integral. While it might be nice to indicate that say Ne(C1, 0) in the
+        # ode solution (even if somewhat obvious due to C1 appearing in a denominator),
+        # the call to solve below won't handle a Piecewise very well (see issue #20269).
+        # For example, the integral evaluation above returns in one case:
+        # Eq(Piecewise((2*atan(sqrt(C1 + 2*exp(f(x)))/sqrt(-C1))/sqrt(-C1), Ne(C1, 0)),
+        #              (-2/sqrt(C1 + 2*exp(f(x))), True)), C2 + x)
+        # Below we look for a Piecewise result with two branches, a Ne(C1, k) for
+        # some is_Number k branch, and an otherwise branch (the otherwise branch
+        # and the condition Ne(C1, k) are ignored).
+        piecewise = [a for a in eq.args if a.is_Piecewise]
+        if len(piecewise) == 1:
+            piecewise_args = piecewise[0].args
+            # Ignore otherwise branch of Piecewise:
+            piecewise_args = [a for a in piecewise_args if a[1] != True]
+            if len(piecewise_args) == 1:
+                ne_branch = piecewise_args[0]
+                ne = ne_branch[1]
+                if isinstance(ne, Ne) and set(ne.args) & constants:
+                    ne = ne.canonical
+                    if ne.rhs.is_Number:
+                        assert ne.lhs in constants
+                        # Keep only the constrained branch (but ignore the Ne condition
+                        # imposed on one of the arbitrary constants):
+                        other_side = next(a for a in eq.args if not a.is_Piecewise)
+                        eq = Eq(ne_branch[0], other_side)
```

Comment:
This change makes sense. The only time it will result in a new expression is when we're dealing with Piecewises of the form we'd like to disregard. It wouldn't change any non-Piecewise expression so it's always safe to do. The only potential problem is if integrate ever returns a condition of the form e.g. Ne(C1, 42) (but we can always add more logic for that case plus I'm not sure how to even craft an ode where an integration result with such a condition is used).